Frontend : Render Mode - Change - Class base render mode

Nevoss opened the pull request.

Nevoss committed:
Frontend : Render Mode - Change - Replace the basic string render mode into class base render mode.

matipojo commented on core/frontend/render-mode-manager.php:
```diff
@@ -0,0 +1,105 @@
+		return $this;
+	}
+
+	private function run_current_actions() {
```
```suggestion
	private function add_current_actions() {
```

matipojo commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,62 @@
+	 *
+	 * @return static
+	 */
+	public function prepare_render() {
+		return $this;
+	}
+
+	/**
+	 * By default do not do nothing.
+	 *
+	 * @return static
+	 */
+	public function enqueue_scripts() {
+		return $this;
+	}
+
+	/**
+	 * By default do not do nothing.
+	 *
+	 * @return static
+	 */
+	public function enqueue_styles() {
+		return $this;
+	}
```
Don't return a value.

matipojo commented on core/frontend/render-modes/render-mode-interface.php:
```diff
@@ -0,0 +1,39 @@
+namespace Elementor\Core\Frontend\RenderModes;
+
+interface Render_Mode_Interface {
+	const QUERY_STRING_PARAM_NAME = 'render_mode';
+	const QUERY_STRING_NONCE_PARAM_NAME = 'render_mode_nonce';
+	const NONCE_ACTION_PATTERN = 'render_mode_{post_id}';
```
Move the the base. (it's not a part of forcing implementation.)

matipojo commented on includes/frontend.php:
```diff
@@ -1156,36 +1154,18 @@ public function create_action_hash( $action, array $settings = [] ) {
-		if ( ! in_array( $render_mode, $available_render_modes, true ) ) {
-			$render_mode = self::RENDER_MODE_NORMAL;
+	public function is_static_render_mode() {
+		if ( ! $this->render_mode_manager ) {
```
Explain.

matipojo commented on tests/phpunit/elementor/includes/base/test-element-base.php:
```diff
@@ -178,20 +178,4 @@ public function test_add_invisible_class_attribute_when_animation_on() {
 		$this->assertTrue( in_array( 'elementor-invisible', $element->get_render_attributes( '_wrapper', 'class' ) ) );
 	}
-
-	public function test_do_not_add_invisible_class_attribute_when_animation_on_and_render_mode_is_static() {
```
Where it moved?

Nevoss replied:
Removed for now, cannot test it because for now there is no render_mode class that is static.

Nevoss committed:
Frontend : Render Mode - Change - Code review.

Nevoss committed:
Frontend : Render Mode - Change - Send the current render mode manager in the action.

matipojo commented on core/frontend/render-mode-manager.php:
```diff
@@ -58,14 +59,14 @@ private function set_current_render_mode() {
 		$key = null;
 		$nonce = null;
-		$nonce_action = str_replace( '{post_id}', $post_id, Render_Mode_Interface::NONCE_ACTION_PATTERN );
+		$nonce_action = str_replace( '{post_id}', $post_id, Render_Mode_Base::NONCE_ACTION_PATTERN );
```
Move to the manager.

matipojo replied:
Use a method to replace here and on `core/frontend/render-modes/render-mode-base.php:24`

matipojo commented on includes/frontend.php:
```diff
@@ -1155,12 +1155,12 @@ public function create_action_hash( $action, array $settings = [] ) {
 	 */
 	public function is_static_render_mode() {
+		// The render mode manager is exists only in frontend,
+		// so bu default if it is not exist the method will return false.
```
```suggestion
		// so by default if it is not exist the method will return false.
```

Nevoss committed:
Frontend : Render Mode - Change - Init the render mode manager before admin bar.

Nevoss committed:
Frontend : Render Mode - Change - Replace const

matipojo commented on core/frontend/render-modes/render-mode-interface.php:
```diff
@@ -0,0 +1,33 @@
+<?php
+namespace Elementor\Core\Frontend\RenderModes;
+
```
Exit if not `ABSPATH`.

Nevoss committed:
Frontend : Render Mode - Change - Typo

Nevoss committed:
Frontend : Render Mode - Fix - ABSPATH + unused import

Nevoss committed:
Merge branch 'release/3.0.0' of github.com:elementor/elementor into feature/render-mode-manager

Nevoss committed:
Frontend : Render Mode - Doc - document some methods.

matipojo commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,57 @@
+			),
+		], get_permalink( $post_id ) );
+	}
+
```
Add a permissions callback.

`Plugin::$instance->documents->get( $post_id )
$document->is_editable_by_current_user()`

matipojo commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,57 @@
+	 * @return string
+	 */
+	public static function get_url( $post_id ) {
+		return add_query_arg( [
```
Move to `Plugin::$instance->...manager->get_base_url( static::get_name() );`

Nevoss committed:
Frontend - Render Mode - Fix - move logic to manager + check for permissions.

Nevoss committed:
Frontend : Render Mode - Fix - call get_post method and not $this->post.

matipojo commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -1,27 +1,41 @@
+	/**
+	 * @var Document
+	 */
+	protected $post;
```
```suggestion
	protected $document;
```

Nevoss committed:
Frontend - Render Mode - enqueue scripts from Render mode + var name change.

Nevoss committed:
Frontend : Render Mode - Test - Render_Mode_Manager test.

Nevoss committed:
Frontend : Render Mode - Test - Render_Mode permissions.

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,94 @@
+	/**
+	 * By default do not do nothing.
+	 */
+	public function enqueue_scripts() {
+		//
+	}
+
+	/**
+	 * By default do not do nothing.
+	 */
+	public function enqueue_styles() {
+		//
+	}
+
```
Why do you need to implement an empty method that has already been implemented in the interface?

Nevoss replied:
In the interface it is only a declaration and not an implementation.

kobizz replied:
The idea of an interface is forcing. If you don't want to force, you shouldn't put it inside the interface.

kobizz commented on core/frontend/render-modes/render-mode-interface.php:
```diff
@@ -0,0 +1,50 @@
+<?php
```
Why do you need an interface if there is already an abstract class?

matipojo replied:
I order to validate the class type without create an instance.
https://github.com/elementor/elementor/pull/12010/files#diff-8e13e40fed5c012a5e93399f9f22a20dR60-R63

kobizz replied:
The problem with interfaces in PHP is that you have to declare all your methods as public, even though they should be protected or private. Additionally, forcing interface implementation is uncommon in our development conventions and needs further discussion.

Nevoss committed:
Frontend : Render Mode - Test - Add assertions for enqueue scripts and styles.

Nevoss committed:
Frontend : Render Mode - Change - Remove interface + test register_render_mode.

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,106 @@
+		return Render_Mode_Manager::get_base_url( $post_id, static::get_name() );
+	}
+
+	/**
+	 * By default do not do nothing.
+	 */
```
It seems like a wrong comment

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,106 @@
+	 * By default do not do nothing.
+	 */
+	public function prepare_render() {
+		add_action( 'wp_enqueue_scripts', function () {
+			$this->enqueue_scripts();
+		}, static::ENQUEUE_SCRIPTS_PRIORITY );
+
+		add_action( 'wp_enqueue_scripts', function () {
+			$this->enqueue_styles();
+		}, static::ENQUEUE_STYLES_PRIORITY );
```
You can use the short syntax: `add_action( 'wp_enqueue_scripts', [ $this, 'enqueue_scripts' ], static::ENQUEUE_SCRIPTS_PRIORITY );`

matipojo replied:
It's for keeping the target function protected.

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	/**
+	 * By default do not do nothing.
```
```suggestion
	 * By default do not do anything.
```

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	/**
+	 * By default do not do nothing.
```
```suggestion
	 * By default do not do anything.
```

kobizz commented on core/frontend/render-modes/render-mode-base.php:
```diff
@@ -0,0 +1,106 @@
+	 *
+	 * @return string
+	 */
+	abstract public static function get_name();
```
PHP throws a strict standard notice when you use an abstract static function

Nevoss committed:
Frontend : Render Mode - Fix - code review.

- Cannot use 'abstract public static function' in php 5.6
- Comment improvements.
